Feat : Add aggregation from kube service endpoints feature in metrics API scaler

julianguinard opened the pull request:
**TL:DR**

Add the ability for metrics-api scaler to get metrics from all endpoint targets of a kubernetes API service (`aggregateFromKubeServiceEndpoints: true` metadata) and aggregate them using `average`, `sum`, `min` or `max` aggregation functions (`aggregationType` metadata), which is a handy feature in an environment where one didn't set up a metric aggregator/scraping stack (i.e prometheus), or simply doesn't want to use their monitoring stack to fetch and serve metrics from customers workload in their own kubernetes clusters, and leave the metrics API's reponsability up to the customer

This PR comes from an issue met when metrics-api scaler targets an internal kubernetes service, [such as the one from metrics_api E2E tests](https://github.com/kedacore/keda/blob/main/tests/scalers/metrics_api/metrics_api_test.go#L106) but, [unlike it](https://github.com/kedacore/keda/blob/main/tests/scalers/metrics_api/metrics_api_test.go#L86), has more than 1 replica serving it, leading to inconsistent HPA average calculation

The more a kubernetes service used by metrics API scaler has pods as targets, the less likely it is for the metrics API scaler to get all metrics from all these pods, leading to inconsistent HPA average metric computation and eventually to scaling issues. Especially in configurations where kubernetes services have "sticky" configurations

Below are screenshots of the modified E2E metrics_api_test.go, commenting out `aggregateFromKubeServiceEndpoints: true` addition to metadata, being executed & failing because scale in never occurs as every metric from all 10 replicas behind metrics server deployment are not all well taken into account

![Screenshot from 2025-02-24 17-06-35](https://github.com/user-attachments/assets/8e517ee0-821d-42c5-a81b-7ddd914ba55b)

logs from keda-operator show same value being returned to metrics-api scaler by the kubernetes service (`16`)
![Screenshot from 2025-02-24 16-18-07](https://github.com/user-attachments/assets/f27450b1-9702-4f29-ba1e-0b66db436709)

As a consequence, test eventually fails because scaled deployment is unable to scale out within 3 minutes timeframe
![Screenshot from 2025-02-24 16-19-00](https://github.com/user-attachments/assets/bd4fedb9-2165-4c76-9800-7e002748c47f)

Uncommenting `aggregateFromKubeServiceEndpoints: true` addition to metadata results in metric aggregation mode and same test passes this time

![Screenshot from 2025-02-24 16-22-14](https://github.com/user-attachments/assets/71282352-6c9b-40ea-bb92-664619b18d94)
![Screenshot from 2025-02-24 16-22-07](https://github.com/user-attachments/assets/31929f73-a0d9-473d-b045-a243fbea1a5d)



### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Helm PR : https://github.com/kedacore/charts/pull/739
Docs PR : https://github.com/kedacore/keda-docs/pull/1541

zroubalik commented:
/run-e2e metrics_api

julianguinard commented:
> /run-e2e metrics_api **Update:** You can check the progress [here](https://github.com/kedacore/keda/actions/runs/13677465647)

@zroubalik thanks for triggering E2E, which ran fine

julianguinard commented:
@zroubalik FYI I added [this commit](https://github.com/kedacore/keda/pull/6565/commits/61ebd1f74a03e14cddc4ac2085f1758e88d92780) to make E2E more resilient on clusters where launching all metrics servers replicas take a longer time than expected to start : we now sleep 1 second & try 4 more times to validate expected pods before failing

It also fixes formatting of each endpoint URL to reach for aggregating metrics (double "/" in path, [did not trigger errors in E2E](https://github.com/kedacore/keda/actions/runs/13677465647/job/38241602315) because server ran by ghcr.io/kedacore/tests-metrics-api seems to cope with it anyway)

JorTurFer commented:
/run-e2e metrics_api

julianguinard commented:
> /run-e2e metrics_api **Update:** You can check the progress [here](https://github.com/kedacore/keda/actions/runs/13866234156)

@JorTurFer thanks for E2E trigger which all ran fine

waiting for reviews & opinions on this now :)

JorTurFer commented on pkg/scalers/metrics_api_scaler.go:
```diff
@@ -408,8 +455,147 @@ func getValueFromYAMLResponse(body []byte, valueLocation string) (float64, error
+	// if we wish to aggregate metric from a kubernetes service then we need to query each endpoint behind the service
+	if s.metadata.aggregateFromKubeServiceEndpoints {
+		endpointsUrls, err := s.getEndpointsUrlsFromServiceURL(ctx, s.metadata.url)
+		if err != nil {
+			s.logger.Error(err, "Failed to get kubernetes endpoints urls from configured service URL. Falling back to querying url configured in metadata")
+		} else {
+			if len(endpointsUrls) == 0 {
+				s.logger.Error(err, "No endpoints URLs were given for the service name. Falling back to querying url configured in metadata")
+			} else {
+				aggregatedMetric, err := s.aggregateMetricsFromMultipleEndpoints(ctx, endpointsUrls)
+				if err != nil {
+					s.logger.Error(err, "No aggregated metrics could be computed from service endpoints. Falling back to querying url configured in metadata")
+				} else {
+					return aggregatedMetric, err
+				}
+			}
+		}
```
I'm not sure if we should fallback here, if a user wants to use the feature and it fails, the fallback can produce unexpected scaling behaviors

julianguinard replied:
yes, I was not really sure about it either, but thought that worst case scenario it would at least behave like the current version of metrics API scaler does and just query the metrics API URL

Would you like me to just return error here instead?

JorTurFer replied:
I'd return an error or, at least, returning an error as default and allowing customizing this via parameter.
I imagine the case where I have 5 instances returning 1000, and suddenly I lose communication with 3 of them, scaling based on 2000 instead of 5000 can produce an unintended scaling in with a service disruption.

In this case, falling back sounds as default risky to me

julianguinard replied:
ok @JorTurFer  I [added this commit](https://github.com/kedacore/keda/pull/6565/commits/25f9b0dda7c553634b22780fff4d89896badb725) in order not to fallback & simply return error whenever encountered

julianguinard replied:
Hello! 
Seems E2E ran fine again. @JorTurFer did you have time to check the updated version fixed according to your remarks by any chance?
thanks! :)

JorTurFer commented:
/run-e2e metrics_api

JorTurFer commented:
/run-e2e metrics_api

JorTurFer commented on pkg/scalers/metrics_api_scaler.go:
```diff
@@ -417,8 +464,140 @@ func getValueFromYAMLResponse(body []byte, valueLocation string) (float64, error
+		namespace := splittedHost[1]
+		podPort := url.Port()
+		// infer port from service scheme when not set explicitly
+		if podPort == "" {
+			if url.Scheme == secureHTTPScheme {
+				podPort = "443"
+			} else {
+				podPort = "80"
+			}
+		}
+		// get service serviceEndpoints
+		serviceEndpoints := &corev1.Endpoints{}
+
+		err := s.kubeClient.Get(ctx, client.ObjectKey{
+			Namespace: namespace,
+			Name:      serviceName,
+		}, serviceEndpoints)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, subset := range serviceEndpoints.Subsets {
+			foundPort := ""
+			for _, port := range subset.Ports {
+				if strconv.Itoa(int(port.Port)) == podPort {
+					foundPort = fmt.Sprintf(":%d", port.Port)
+					break
+				}
+			}
+			if foundPort == "" {
+				s.logger.Info(fmt.Sprintf("Warning : could not find port %s in endpoint slice for service %s.%s definition. Will infer port from %s scheme", podPort, serviceName, namespace, url.Scheme))
+			}
+			for _, address := range subset.Addresses {
+				if address.NodeName != nil {
+					endpointUrls = append(endpointUrls, fmt.Sprintf("%s://%s%s%s", url.Scheme, address.IP, foundPort, url.Path))
+				}
+			}
+		}
+	}
+	return endpointUrls, err
```
let's remove indentation
```suggestion
	url, err := neturl.Parse(serviceURL)
	if err != nil {
		s.logger.Error(err, "Failed parsing url for metrics API")
		return nil, err
	} 
	
	splittedHost := strings.Split(url.Host, ".")
	if len(splittedHost) < 2 {
		return nil, fmt.Errorf("invalid hostname %s : expected at least 2 elements, first being service name and second being the namespace", url.Host)
	}
	serviceName := splittedHost[0]
	namespace := splittedHost[1]
	podPort := url.Port()
	// infer port from service scheme when not set explicitly
	if podPort == "" {
		if url.Scheme == secureHTTPScheme {
			podPort = "443"
		} else {
			podPort = "80"
		}
	}
	// get service serviceEndpoints
	serviceEndpoints := &corev1.Endpoints{}

	err := s.kubeClient.Get(ctx, client.ObjectKey{
		Namespace: namespace,
		Name:      serviceName,
	}, serviceEndpoints)
	if err != nil {
		return nil, err
	}

	for _, subset := range serviceEndpoints.Subsets {
		foundPort := ""
		for _, port := range subset.Ports {
			if strconv.Itoa(int(port.Port)) == podPort {
				foundPort = fmt.Sprintf(":%d", port.Port)
				break
			}
		}
		if foundPort == "" {
			s.logger.Info(fmt.Sprintf("Warning : could not find port %s in endpoint slice for service %s.%s definition. Will infer port from %s scheme", podPort, serviceName, namespace, url.Scheme))
		}
		for _, address := range subset.Addresses {
			if address.NodeName != nil {
				endpointUrls = append(endpointUrls, fmt.Sprintf("%s://%s%s%s", url.Scheme, address.IP, foundPort, url.Path))
			}
		}
	}
	
	return endpointUrls, err
```

JorTurFer commented on pkg/scalers/metrics_api_scaler.go:
```diff
@@ -417,8 +464,140 @@ func getValueFromYAMLResponse(body []byte, valueLocation string) (float64, error
+	const maxGoroutines = 5
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+	sem := make(chan struct{}, maxGoroutines)
```
Let's use a semaphore directly instead of a channel -> https://pkg.go.dev/golang.org/x/sync/semaphore

semaphore is part of standard lib

JorTurFer reviewed (commented):
Nice job! I've kept some implementation ideas inline

julianguinard commented:
@JorTurFer I added your implementation suggestions & also fixed a few conflicts following latest release.

Could you please-re-review & run E2E?

Thanks!

JorTurFer commented:
/run-e2e metrics_api

rickbrouwer commented:
/run-e2e metrics_api

JorTurFer commented:
@kedacore/keda-core-contributors PTAL

julianguinard commented on pkg/scalers/metrics_api_scaler.go:
```diff
@@ -421,8 +473,145 @@ func getValueFromYAMLResponse(body []byte, valueLocation string) (float64, error
+			metric, err := s.getMetricValueFromURL(ctx, &endpointURL)
+
+			if err != nil {
+				s.logger.Info(fmt.Sprintf("Error fetching metric for %s: %v\n", url, err))
```
done in https://github.com/kedacore/keda/pull/6565/commits/8b737de39d5b71781224c4557665ea999e165d9a

julianguinard commented on pkg/scalers/metrics_api_scaler.go:
```diff
@@ -421,8 +473,145 @@ func getValueFromYAMLResponse(body []byte, valueLocation string) (float64, error
+		}(endpointURL)
+	}
+
+	wg.Wait()
```
right my bad, this waitgroup logic was forgotten dead code from https://github.com/kedacore/keda/pull/6565/commits/afa47cbee924e674d6e64ad0fd9c2fd0784ea468 that I forgot to remove.
Semaphore acquisition was indeed added [here](https://github.com/kedacore/keda/pull/6565/commits/afa47cbee924e674d6e64ad0fd9c2fd0784ea468#diff-c832e36c124014f31496fdbc15f77397f48b73855a7b7441fd95525ed6112f54R559) in the same commit

julianguinard replied:
hello @JorTurFer I hope your are well, happy new year to you!
I realized post-merge that I just removed the waitgroup following your comment but forgot to add the final semaphore acquisition after the loop, fixed in https://github.com/kedacore/keda/pull/7370

zroubalik commented on config/rbac/role.yaml:
```diff
@@ -9,6 +9,7 @@ rules:
   resources:
   - configmaps
   - configmaps/status
+  - endpoints
```
We should use EndpointSlices:
https://kubernetes.io/blog/2025/04/24/endpoints-deprecation/

julianguinard replied:
@zroubalik switch to end point slices api done

You may review/rerun E2E whenever wanted

wozniakjan reviewed (commented):
this looks super usefull, few items regarding EndpointSlice usage below

wozniakjan commented on pkg/scalers/metrics_api_scaler.go:
```diff
@@ -278,8 +300,156 @@ func getValueFromYAMLResponse(body []byte, valueLocation string) (float64, error
+		for _, eps := range endpointSlice.Endpoints {
+			// as suggested in https://github.com/kedacore/keda/pull/6565#discussion_r2395073047, make sure we take endpoint into account
+			// only when it's ready
+			if eps.Conditions.Ready == nil || !*eps.Conditions.Ready {
```
it is a bit counterintuitive but `Conditions.Ready == nil` should be treated as "ready", see also [api note](https://github.com/kubernetes/api/blob/c397f90a55bd51f3f33e729ae6728c219beab2cf/discovery/v1/types.go#L137-L144)

```suggestion
			if eps.Conditions.Ready != nil && !*eps.Conditions.Ready {
```

julianguinard commented:
E2E all green. @wozniakjan feel free to have a thidr look at it whenever you wish :)
